libraries/chain_kv/include/b1/session/shared_bytes.hpp

```diff
@@ -0,0 +1,292 @@
+#pragma once
+
+#include <algorithm>
+#include <string_view>
+
+#include <boost/pool/pool.hpp>
+
+#include <b1/session/shared_bytes_fwd_decl.hpp>
+#include <b1/session/key_value_fwd_decl.hpp>
+
+namespace eosio::session {
+
+// \brief An immutable type to represent a pointer and a length.
+//
+class shared_bytes final {
+public:
+    friend class key_value;
+    
+    template <typename T, typename Allocator>
+    friend shared_bytes make_shared_bytes(const T* data, size_t length, Allocator& a);
+
+    template <typename Allocator>
+    friend shared_bytes make_shared_bytes(const void* data, size_t length, Allocator& a);
+    
+    template <typename T>
+    friend shared_bytes make_shared_bytes(const T* data, size_t length);
+    
+    friend key_value make_kv(shared_bytes key, shared_bytes value);
+
+    template <typename Key, typename Value, typename Allocator>
+    friend key_value make_kv(const Key* the_key, size_t key_length, const Value* the_value, size_t value_length, Allocator& a);
+    
+    template <typename Key, typename Value>
+    friend key_value make_kv(const Key* the_key, size_t key_length, const Value* the_value, size_t value_length);
+
+    template <typename Allocator>
+    friend key_value make_kv(const void* key, size_t key_length, const void* value, size_t value_length, Allocator& a);
+    
+    friend key_value make_kv(const void* key, size_t key_length, const void* value, size_t value_length);
+
+public:
+    shared_bytes(const shared_bytes& b);
+    shared_bytes(shared_bytes&& b);
+    ~shared_bytes();
+    
+    shared_bytes& operator=(const shared_bytes& b);
+    shared_bytes& operator=(shared_bytes&& b);
+    
+    const void * const data() const;
+    size_t length() const;
+    
+    bool operator==(const shared_bytes& other) const;
+    bool operator!=(const shared_bytes& other) const;
+    
+    static const shared_bytes invalid;
+    
+private:
+    shared_bytes() = default;
+    
+private:
+    // To allow for sharing a chunk of memory across multiple shared_byte instances, we need to keep track
+    // of the chunk start in each instance to pass back to the allocator.
+    uint64_t* m_chunk_start{nullptr};
+    // The encoded address of the memory allocator that instantiated the memory this instance points to.
+    // \warning The memory allocator MUST stay in scope during the lifetime of this instance.
+    uint64_t* m_memory_allocator_address{nullptr};
+    // \brief A pointer to a counter to keep track of the number of shared_bytes instances that point to the memory address.
+    // \remark When the counter reaches 0, the last instance will free the memory by invoking the free method on the memory allocator.
+    // \warning Assuming that this type will be accessed synchronously.
+    size_t* m_use_count_address{nullptr};
```

Comment:
This comment implies that this refcount makes the system not thread safe.  is there a reason to use a raw scalar vs an atomic type counter?

Reply:
I was under the impression that the system operated synchronously.  So given that, it didn't seem necessary to have the overhead of an atomic.  Is the system advertised as thread safe?  I can make it atomic if so.

Reply:
@b1bart is your point that this could eventually be used asynchronously? If we want this code to be safe to access asynchronously, we first need to decide the use cases (i.e. what is going to be used asynchronously) and then make that usage safe.  I see this comment as a bread crumb for if this ever happens, but really shouldn't matter, since we will need to look at the whole design to make sure it is safe, but making this an atomic seems to me to be a lot of cost compared to the amount of work that will still be needed if we do make it asynchronous.

Reply:
I don't have a clear view of the whole system, and while it is true the db system operates synchronously, plenty of things that inspect data during operation operate asynchronously.  My concern is centered around the basic notion that a a future engineer may not have the nuance view to not place one of these spans in an action_trace etc.  

something like:
```
struct shared_span {
   size_t                     m_length = 0;
   std::shared_ptr<uint8_t[]> m_data;
}
```

seems conceptually simpler, is thread safe, and still supports safely aliasing within a span via `std::shared_ptr`'s aliasing constructor